src/D2L.CodeStyle.Analyzers/Language/EscapeNonAsciiCharsInLiteralsAnalyzer.cs

```diff
@@ -0,0 +1,200 @@
+﻿using System.Collections.Immutable;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+	// We are doing this because we (unfortunately) have a mix of encoding for
+	// our source code files, and the encoding of the file impacts how strings
+	// are interpreted. We really ought to clean that up, but until that
+	// happens we can be safer if we avoid non-ASCII characters in our
+	// literals. Automated refactoring has bit us before with this.
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class EscapeNonAsciiCharsInLiteralsAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+			=> ImmutableArray.Create( Diagnostics.EscapeNonAsciiCharsInLiteral );
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+
+			context.RegisterSyntaxNodeAction(
+				CheckForUnicodeLiterals,
+				SyntaxKind.StringLiteralExpression
+			);
+
+			context.RegisterSyntaxNodeAction(
+				CheckForUnicodeLiterals,
+				SyntaxKind.CharacterLiteralExpression
+			);
+		}
+
+		private static void CheckForUnicodeLiterals(
+			SyntaxNodeAnalysisContext ctx
+		) {
+			// the node could be something like:
+			//   "foo"
+			//   @"foo"
+			//   'x'
+			var literalExpr = (LiteralExpressionSyntax)ctx.Node;
+
+			// We can't handle verbatim strings for the same reason as these
+			// folks: https://github.com/dotnet/codeformatter/issues/39 (TODO)
+			if ( literalExpr.Token.IsVerbatimStringLiteral() ) {
+				return;
+			}
+
+			string token = literalExpr.Token.Text;
+			string escapedToken = null;
+
+			if( StrictlyAscii( token, out escapedToken ) ) {
+				return;
+			}
+
+			bool isChar =
+				literalExpr.Kind() == SyntaxKind.CharacterLiteralExpression;
+
+			var fixProps = ImmutableDictionary.CreateBuilder<string, string>();
+			fixProps[EscapeCharCodeFix.ESCAPED] = escapedToken;
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.EscapeNonAsciiCharsInLiteral,
+					literalExpr.GetLocation(),
+					fixProps.ToImmutable(),
+					isChar ? "char" : "string",
+					escapedToken
+				)
+			);
+		}
+
+		private static bool StrictlyAscii(
+			string token,
+			out string escapedToken
+		) {
+			var sb = new StringBuilder();
+			var copyStartIdx = 0;
+
+			// invariant: copyStartIdx < idx
+			// Note: the enclosing quotes are included in val; don't bother looking at them.
+			for( int idx = 1; idx < token.Length - 1; idx++ ) {
+				if( token[idx] < 0x80 ) {
+					continue;
+				}
+
+				// copy all the ascii chars we've seen between the last copy
+				// and now (not inclusive) into sb
+				sb.Append( token, copyStartIdx, idx - copyStartIdx );

```

Comment:
It's "basically a no-op" when the string is ASCII. Microsoft did it with two passes (one to detect nonASCIIness, one to do the copying) which did char-by-char copying in the second step. It's technically more work to do that (two passes, and char by char is still worse than chunks at a time) but its so neglible and only happens during errors anyway so really not interesting.

Honestly I just wrote it this way because it came out of my head this way. I'll look at it again.